tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	defer t.mu.RUnlock()
+	for _, host := range t.hosts {
+		host.decay()
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) decayLoop() {
+	ticker := time.NewTicker(t.decayPeriod / bucketCount)
+	for {
+		select {
+		case <-ticker.C:
+			t.decay()
+		case <-t.stopChan:
+			return
+		}
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) Init(s *Session) {
+	t.getKeyspaceMetadata = s.KeyspaceMetadata
+	go t.decayLoop()
+}
+
+func (t *TokenAwareBanditHostPolicy) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.stopChan)
+	})
+}
+
+// hostInfoSlice converts t.hosts to a slice of *HostInfo.
+// It must be called with t.mu held for reading.
+func (t *TokenAwareBanditHostPolicy) hostInfoSlice() []*HostInfo {
+	hosts := make([]*HostInfo, 0, len(t.hosts))
+	for _, host := range t.hosts {
+		host.mu.RLock()
+		hostInfo := host.hostInfo
+		host.mu.RUnlock()
+		hosts = append(hosts, hostInfo)
+	}
+	return hosts
+}
+
+func (t *TokenAwareBanditHostPolicy) IsLocal(host *HostInfo) bool {
+	return t.localDatacenter != "" && host.DataCenter() == t.localDatacenter
+}
+
+func (t *TokenAwareBanditHostPolicy) KeyspaceChanged(update KeyspaceUpdateEvent) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.updateReplicas(update.Keyspace)
+}
+
+// updateReplicas initiates updating of replicas for the given keyspace.
+// It must be called while the caller holds t.mu mutex locked for writes.
+func (t *TokenAwareBanditHostPolicy) updateReplicas(keyspace string) {
+	ks, ok := t.keyspaces[keyspace]
+	if !ok {
+		ks = &tokenAwareBanditKeyspace{}
+		t.keyspaces[keyspace] = ks
+	}
+
+	ks.needsUpdate = true
+
+	if ks.updating {
+		return
+	}
+
+	ks.updating = true
+	go t.updateReplicasLoop(ks, keyspace)
+}
+
+// replicasNeedUpdate checks if the keyspace needs updating.
+// If the keyspace does not need to be updated, it resets the updating flag.
+func (t *TokenAwareBanditHostPolicy) replicasNeedUpdate(ks *tokenAwareBanditKeyspace) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if ks.needsUpdate {
+		ks.needsUpdate = false
+		return true
+	}
+	ks.updating = false
+	return false
+}
+
+// updateReplicasLoop performs actual update of replicas for a keyspace.
+// updateReplicasLoop must be called in a goroutine not holding the t.mu mutex.
+func (t *TokenAwareBanditHostPolicy) updateReplicasLoop(ks *tokenAwareBanditKeyspace, keyspaceName string) {
+	for t.replicasNeedUpdate(ks) {
+		// We can't call getKeyspaceMetadata while holding the lock since getKeyspaceMetadata can do I/O.
+		metadata, err := t.getKeyspaceMetadata(keyspaceName)
+		if err != nil {
+			t.logger.Println(err)
+			continue
+		}
+
+		strategy := getStrategy(metadata)
+		var replicas tokenRingReplicas
+
+		if strategy != nil {
+			t.mu.RLock()
+			tokenRing := t.tokenRing
+			t.mu.RUnlock()
+			replicas = strategy.replicaMap(tokenRing)
+		}
+
+		t.mu.Lock()
+		ks.replicas = replicas
+		t.mu.Unlock()
+	}
+	if t.afterReplicasUpdated != nil {
+		t.afterReplicasUpdated(keyspaceName)
+	}
+}
+
+// SetPartitioner is called to set the partitioner used.
```

Comment:
```suggestion
// SetPartitioner is called to set the partitioner used to map partition keys to token.
// It also resets something and it will be great to explain
```

Let's make docs more descriptive plz 🙏

Reply:
Updated docs / added inline comments inside methods.